File: 1105-filling-bookcase-shelves/1105-filling-bookcase-shelves.py
class Solution:
    def minHeightShelves(self, b: List[List[int]], sw: int) -> int:
        dp = [0] + [float('inf')] * len(b)

        for i in range(1, len(dp)):
            currh = currw = 0
            for j in range(i - 1, -1, -1):
                currw += b[j][0]
                if currw > sw: 
                    break
                currh = max(currh, b[j][1])
                dp[i] = min(dp[i], dp[j] + currh)

        return dp[-1

File: 1105-filling-bookcase-shelves/1105-filling-bookcase-shelves.cpp
class Solution {
public:
    int minHeightShelves(vector<vector<int>>& b, int sw) {
        vector <int> dp(b.size() + 1, INT_MAX);
        dp[0] = 0;

        for (int i = 1; i <= b.size(); i++) {
            int ch = 0, cw = 0;
            for (int j = i - 1; j >= 0; j--) {
                cw += b[j][0];
                if (cw > sw) break;
                ch = max(ch, b[j][1]);
                dp[i] = min(dp[i], dp[j] + ch);
            }
        }

        return dp[b.size()];
    }
};